File: rb/Valorant_sdk.rb
# Valorant SDK

require_relative 'utility/struct/voxgig_struct'
require_relative 'core/utility_type'
require_relative 'core/spec'
require_relative 'core/helpers'

# Load utility registration
require_relative 'utility/register'

# Load config and features
require_relative 'config'
require_relative 'feature/base_feature'
require_relative 'features'

# Load typed models (Struct value objects).
require_relative 'Valorant_types'


class ValorantSDK
  attr_accessor :mode, :features, :options

  def initialize(options = {})
    @mode = "live"
    @features = []
    @options = nil

    utility = ValorantUtility.new
    @_utility = utility

    config = ValorantConfig.make_config

    @_rootctx = utility.make_context.call({
      "client" => self,
      "utility" => utility,
      "config" => config,
      "options" => options || {},
      "shared" => {},
    }, nil)

    @options = utility.make_options.call(@_rootctx)

    if VoxgigStruct.getpath(@options, "feature.test.active") == true
      @mode = "test"
    end

    @_rootctx.options = @options

    # Add features from config.
    feature_opts = ValorantHelpers.to_map(VoxgigStruct.getprop(@options, "feature"))
    if feature_opts
      items = VoxgigStruct.items(feature_opts)
      if items
        items.each do |item|
          fname = item[0]
          fopts = ValorantHelpers.to_map(item[1])
          if fopts && fopts["active"] == true
            utility.feature_add.call(@_rootctx, ValorantFeatures.make_feature(fname))
          end
        end
      end
    end

    # Add extension features.
    extend_val = VoxgigStruct.getprop(@options, "extend")
    if extend_val.is_a?(Array)
      extend_val.each do |f|
        if f.respond_to?(:get_name)
          utility.feature_add.call(@_rootctx, f)
        end
      end
    end

    # Initialize features.
    @features.each do |f|
      utility.feature_init.call(@_rootctx, f)
    end

    utility.feature_hook.call(@_rootctx, "PostConstruct")
  end

  def options_map
    out = VoxgigStruct.clone(@options)
    out.is_a?(Hash) ? out : {}
  end

  def get_utility
    ValorantUtility.copy(@_utility)
  end

  def get_root_ctx
    @_rootctx
  end

  def prepare(fetchargs = {})
    utility = @_utility
    fetchargs ||= {}

    ctrl = ValorantHelpers.to_map(VoxgigStruct.getprop(fetchargs, "ctrl")) || {}

    ctx = utility.make_context.call({
      "opname" => "prepare",
      "ctrl" => ctrl,
    }, @_rootctx)

    opts = @options
    path = VoxgigStruct.getprop(fetchargs, "path") || ""
    path = "" unless path.is_a?(String)
    method_val = VoxgigStruct.getprop(fetchargs, "method") || "GET"
    method_val = "GET" unless method_val.is_a?(String)
    params = ValorantHelpers.to_map(VoxgigStruct.getprop(fetchargs, "params")) || {}
    query = ValorantHelpers.to_map(VoxgigStruct.getprop(fetchargs, "query")) || {}
    headers = utility.prepare_headers.call(ctx)

    base = VoxgigStruct.getprop(opts, "base") || ""
    base = "" unless base.is_a?(String)
    prefix = VoxgigStruct.getprop(opts, "prefix") || ""
    prefix = "" unless prefix.is_a?(String)
    suffix = VoxgigStruct.getprop(opts, "suffix") || ""
    suffix = "" unless suffix.is_a?(String)

    ctx.spec = ValorantSpec.new({
      "base" => base, "prefix" => prefix, "suffix" => suffix,
      "path" => path, "method" => method_val,
      "params" => params, "query" => query, "headers" => headers,
      "body" => VoxgigStruct.getprop(fetchargs, "body"),
      "step" => "start",
    })

    # Merge user-provided headers.
    uh = VoxgigStruct.getprop(fetchargs, "headers")
    if uh.is_a?(Hash)
      uh.each { |k, v| ctx.spec.headers[k] = v }
    end

    _, err = utility.prepare_auth.call(ctx)
    raise err if err

    utility.make_fetch_def.call(ctx)
  end

  def direct(fetchargs = {})
    utility = @_utility

    # direct() is the raw-HTTP escape hatch: it always returns a result hash
    # ({ "ok" => ..., ... }) and never raises. prepare() raises on error, so
    # trap that and surface it in the hash.
    begin
      fetchdef = prepare(fetchargs)
    rescue ValorantError => err
      return { "ok" => false, "err" => err }
    end

    fetchargs ||= {}
    ctrl = ValorantHelpers.to_map(VoxgigStruct.getprop(fetchargs, "ctrl")) || {}

    ctx = utility.make_context.call({
      "opname" => "direct",
      "ctrl" => ctrl,
    }, @_rootctx)

    url = fetchdef["url"] || ""
    fetched, fetch_err = utility.fetcher.call(ctx, url, fetchdef)

    return { "ok" => false, "err" => fetch_err } if fetch_err

    if fetched.nil?
      return {
        "ok" => false,
        "err" => ctx.make_error("direct_no_response", "response: undefined"),
      }
    end

    if fetched.is_a?(Hash)
      status = ValorantHelpers.to_int(VoxgigStruct.getprop(fetched, "status"))
      headers = VoxgigStruct.getprop(fetched, "headers") || {}

      # No-body responses (204, 304) and explicit zero content-length must
      # skip JSON parsing — calling json() on an empty body errors.
      content_length = headers.is_a?(Hash) ? headers["content-length"] : nil
      no_body = status == 204 || status == 304 || content_length.to_s == "0"

      json_data = nil
      unless no_body
        jf = VoxgigStruct.getprop(fetched, "json")
        if jf.is_a?(Proc)
          begin
            json_data = jf.call
          rescue StandardError
            # Non-JSON body — leave data nil, keep status/headers.
            json_data = nil
          end
        end
      end

      return {
        "ok" => status >= 200 && status < 300,
        "status" => status,
        "headers" => headers,
        "data" => json_data,
      }
    end

    return {
      "ok" => false,
      "err" => ctx.make_error("direct_invalid", "invalid response type"),
    }
  end


  # Idiomatic facade: client.agent.list / client.agent.load({ "id" => ... })
  def agent
    require_relative 'entity/agent_entity'
    @agent ||= AgentEntity.new(self, nil)
  end

  # Deprecated: use client.agent instead.
  def Agent(data = nil)
    require_relative 'entity/agent_entity'
    AgentEntity.new(self, data)
  end


  # Idiomatic facade: client.competitive.list / client.competitive.load({ "id" => ... })
  def competitive
    require_relative 'entity/competitive_entity'
    @competitive ||= CompetitiveEntity.new(self, nil)
  end

  # Deprecated: use client.competitive instead.
  def Competitive(data = nil)
    require_relative 'entity/competitive_entity'
    CompetitiveEntity.new(self, data)
  end


  # Idiomatic facade: client.cosmetic.list / client.cosmetic.load({ "id" => ... })
  def cosmetic
    require_relative 'entity/cosmetic_entity'
    @cosmetic ||= CosmeticEntity.new(self, nil)
  end

  # Deprecated: use client.cosmetic instead.
  def Cosmetic(data = nil)
    require_relative 'entity/cosmetic_entity'
    CosmeticEntity.new(self, data)
  end


  # Idiomatic facade: client.game_mode.list / client.game_mode.load({ "id" => ... })
  def game_mode
    require_relative 'entity/game_mode_entity'
    @game_mode ||= GameModeEntity.new(self, nil)
  end

  # Deprecated: use client.game_mode instead.
  def GameMode(data = nil)
    require_relative 'entity/game_mode_entity'
    GameModeEntity.new(self, data)
  end


  # Idiomatic facade: client.map.list / client.map.load({ "id" => ... })
  def map
    require_relative 'entity/map_entity'
    @map ||= MapEntity.new(self, nil)
  end

  # Deprecated: use client.map instead.
  def Map(data = nil)
    require_relative 'entity/map_entity'
    MapEntity.new(self, data)
  end


  # Idiomatic facade: client.weapon.list / client.weapon.load({ "id" => ... })
  def weapon
    require_relative 'entity/weapon_entity'
    @weapon ||= WeaponEntity.new(self, nil)
  end

  # Deprecated: use client.weapon instead.
  def Weapon(data = nil)
    require_relative 'entity/weapon_entity'
    WeaponEntity.new(self, data)
  end



  def self.test(testopts = nil, sdkopts = nil)
    sdkopts = sdkopts || {}
    sdkopts = VoxgigStruct.clone(sdkopts)
    sdkopts = {} unless sdkopts.is_a?(Hash)

    testopts = testopts || {}
    testopts = VoxgigStruct.clone(testopts)
    testopts = {} unless testopts.is_a?(Hash)
    testopts["active"] = true

    VoxgigStruct.setpath(sdkopts, "feature.test", testopts)

    sdk = ValorantSDK.new(sdkopts)
    sdk.mode = "test"
    sdk
  end
end

File: lua/valorant_sdk.lua
-- Valorant SDK

local vs = require("utility.struct.struct")
local Utility = require("core.utility_type")
local Spec = require("core.spec")
local helpers = require("core.helpers")

-- Load utility registration (populates Utility._registrar)
require("utility.register")

-- Load features
local BaseFeature = require("feature.base_feature")
local features_factory = require("features")


local ValorantSDK = {}
ValorantSDK.__index = ValorantSDK


local function _make_feature(name)
  local factory = features_factory[name]
  if factory ~= nil then
    return factory()
  end
  return features_factory.base()
end

ValorantSDK._make_feature = _make_feature


function ValorantSDK.new(options)
  local self = setmetatable({}, ValorantSDK)
  self.mode = "live"
  self.features = {}
  self.options = nil

  local utility = Utility.new()
  self._utility = utility

  local config = require("config")()

  self._rootctx = utility.make_context({
    client = self,
    utility = utility,
    config = config,
    options = options or {},
    shared = {},
  }, nil)

  self.options = utility.make_options(self._rootctx)

  if vs.getpath(self.options, "feature.test.active") == true then
    self.mode = "test"
  end

  self._rootctx.options = self.options

  -- Add features from config.
  local feature_opts = helpers.to_map(vs.getprop(self.options, "feature"))
  if feature_opts ~= nil then
    local feature_items = vs.items(feature_opts)
    if feature_items ~= nil then
      for _, item in ipairs(feature_items) do
        local fname = item[1]
        local fopts = helpers.to_map(item[2])
        if fopts ~= nil and fopts["active"] == true then
          utility.feature_add(self._rootctx, _make_feature(fname))
        end
      end
    end
  end

  -- Add extension features.
  local extend = vs.getprop(self.options, "extend")
  if type(extend) == "table" then
    for _, f in ipairs(extend) do
      if type(f) == "table" and type(f.get_name) == "function" then
        utility.feature_add(self._rootctx, f)
      end
    end
  end

  -- Initialize features.
  for _, f in ipairs(self.features) do
    utility.feature_init(self._rootctx, f)
  end

  utility.feature_hook(self._rootctx, "PostConstruct")

  -- #BuildFeatures

  return self
end


function ValorantSDK:options_map()
  local out = vs.clone(self.options)
  if type(out) == "table" then
    return out
  end
  return {}
end


function ValorantSDK:get_utility()
  return Utility.copy(self._utility)
end


function ValorantSDK:get_root_ctx()
  return self._rootctx
end


function ValorantSDK:prepare(fetchargs)
  local utility = self._utility

  fetchargs = fetchargs or {}

  local ctrl = helpers.to_map(vs.getprop(fetchargs, "ctrl")) or {}

  local ctx = utility.make_context({
    opname = "prepare",
    ctrl = ctrl,
  }, self._rootctx)

  local options = self.options

  local path = vs.getprop(fetchargs, "path") or ""
  if type(path) ~= "string" then path = "" end

  local method = vs.getprop(fetchargs, "method") or "GET"
  if type(method) ~= "string" then method = "GET" end

  local params = helpers.to_map(vs.getprop(fetchargs, "params")) or {}
  local query = helpers.to_map(vs.getprop(fetchargs, "query")) or {}

  local headers = utility.prepare_headers(ctx)

  local base = vs.getprop(options, "base") or ""
  if type(base) ~= "string" then base = "" end
  local prefix = vs.getprop(options, "prefix") or ""
  if type(prefix) ~= "string" then prefix = "" end
  local suffix = vs.getprop(options, "suffix") or ""
  if type(suffix) ~= "string" then suffix = "" end

  ctx.spec = Spec.new({
    base = base,
    prefix = prefix,
    suffix = suffix,
    path = path,
    method = method,
    params = params,
    query = query,
    headers = headers,
    body = vs.getprop(fetchargs, "body"),
    step = "start",
  })

  -- Merge user-provided headers.
  local uh = vs.getprop(fetchargs, "headers")
  if type(uh) == "table" then
    for k, v in pairs(uh) do
      ctx.spec.headers[k] = v
    end
  end

  local _, err = utility.prepare_auth(ctx)
  if err ~= nil then
    return nil, err
  end

  return utility.make_fetch_def(ctx)
end


function ValorantSDK:direct(fetchargs)
  local utility = self._utility

  local fetchdef, err = self:prepare(fetchargs)
  if err ~= nil then
    return { ok = false, err = err }, nil
  end

  fetchargs = fetchargs or {}
  local ctrl = helpers.to_map(vs.getprop(fetchargs, "ctrl")) or {}

  local ctx = utility.make_context({
    opname = "direct",
    ctrl = ctrl,
  }, self._rootctx)

  local url = fetchdef["url"] or ""
  local fetched, fetch_err = utility.fetcher(ctx, url, fetchdef)

  if fetch_err ~= nil then
    return { ok = false, err = fetch_err }, nil
  end

  if fetched == nil then
    return {
      ok = false,
      err = ctx:make_error("direct_no_response", "response: undefined"),
    }, nil
  end

  if type(fetched) == "table" then
    local status = helpers.to_int(vs.getprop(fetched, "status"))
    local headers = vs.getprop(fetched, "headers") or {}

    -- No-body responses (204, 304) and explicit zero content-length
    -- must skip JSON parsing — calling json() on an empty body errors.
    local content_length = nil
    if type(headers) == "table" then
      content_length = headers["content-length"]
    end
    local no_body = status == 204 or status == 304 or tostring(content_length) == "0"

    local json_data = nil
    if not no_body then
      local jf = vs.getprop(fetched, "json")
      if type(jf) == "function" then
        local ok, result = pcall(jf)
        if ok then
          json_data = result
        end
        -- Non-JSON body: json_data stays nil, status/headers preserved.
      end
    end

    return {
      ok = status >= 200 and status < 300,
      status = status,
      headers = headers,
      data = json_data,
    }, nil
  end

  return {
    ok = false,
    err = ctx:make_error("direct_invalid", "invalid response type"),
  }, nil
end



-- Idiomatic facade: client:agent():list() / client:agent():load({ id = ... })
function ValorantSDK:agent(data)
  local EntityMod = require("entity.agent_entity")
  if data == nil then
    if self._agent == nil then
      self._agent = EntityMod.new(self, nil)
    end
    return self._agent
  end
  return EntityMod.new(self, data)
end

-- Deprecated: use client:agent() instead.
function ValorantSDK:Agent(data)
  local EntityMod = require("entity.agent_entity")
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:competitive():list() / client:competitive():load({ id = ... })
function ValorantSDK:competitive(data)
  local EntityMod = require("entity.competitive_entity")
  if data == nil then
    if self._competitive == nil then
      self._competitive = EntityMod.new(self, nil)
    end
    return self._competitive
  end
  return EntityMod.new(self, data)
end

-- Deprecated: use client:competitive() instead.
function ValorantSDK:Competitive(data)
  local EntityMod = require("entity.competitive_entity")
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:cosmetic():list() / client:cosmetic():load({ id = ... })
function ValorantSDK:cosmetic(data)
  local EntityMod = require("entity.cosmetic_entity")
  if data == nil then
    if self._cosmetic == nil then
      self._cosmetic = EntityMod.new(self, nil)
    end
    return self._cosmetic
  end
  return EntityMod.new(self, data)
end

-- Deprecated: use client:cosmetic() instead.
function ValorantSDK:Cosmetic(data)
  local EntityMod = require("entity.cosmetic_entity")
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:game_mode():list() / client:game_mode():load({ id = ... })
function ValorantSDK:game_mode(data)
  local EntityMod = require("entity.game_mode_entity")
  if data == nil then
    if self._game_mode == nil then
      self._game_mode = EntityMod.new(self, nil)
    end
    return self._game_mode
  end
  return EntityMod.new(self, data)
end

-- Deprecated: use client:game_mode() instead.
function ValorantSDK:GameMode(data)
  local EntityMod = require("entity.game_mode_entity")
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:map():list() / client:map():load({ id = ... })
function ValorantSDK:map(data)
  local EntityMod = require("entity.map_entity")
  if data == nil then
    if self._map == nil then
      self._map = EntityMod.new(self, nil)
    end
    return self._map
  end
  return EntityMod.new(self, data)
end

-- Deprecated: use client:map() instead.
function ValorantSDK:Map(data)
  local EntityMod = require("entity.map_entity")
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:weapon():list() / client:weapon():load({ id = ... })
function ValorantSDK:weapon(data)
  local EntityMod = require("entity.weapon_entity")
  if data == nil then
    if self._weapon == nil then
      self._weapon = EntityMod.new(self, nil)
    end
    return self._weapon
  end
  return EntityMod.new(self, data)
end

-- Deprecated: use client:weapon() instead.
function ValorantSDK:Weapon(data)
  local EntityMod = require("entity.weapon_entity")
  return EntityMod.new(self, data)
end




function ValorantSDK.test(testopts, sdkopts)
  sdkopts = sdkopts or {}
  sdkopts = vs.clone(sdkopts)
  if type(sdkopts) ~= "table" then
    sdkopts = {}
  end

  testopts = testopts or {}
  testopts = vs.clone(testopts)
  if type(testopts) ~= "table" then
    testopts = {}
  end
  testopts["active"] = true

  vs.setpath(sdkopts, "feature.test", testopts)

  local sdk = ValorantSDK.new(sdkopts)
  sdk.mode = "test"

  return sdk
end


return ValorantSDK
